File: android/app/src/test/java/com/antlib/hingewave/CoreTests.kt
package com.antlib.hingewave

import com.antlib.hingewave.core.DetentDetector
import com.antlib.hingewave.core.EffectConfig
import com.antlib.hingewave.core.PhoneMapping
import com.antlib.hingewave.core.Spring
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File
import kotlin.math.abs

class EffectConfigParityTest {
    @Test
    fun kotlinDefaultsMatchEffectJson() {
        val fromFile = EffectConfig.fromJson(File(CorePaths.core, "effect.json").readText())
        assertEquals("core/effect.json and EffectConfig.DEFAULTS differ; update both together", EffectConfig.DEFAULTS, fromFile)
    }
}

class SpringTest {
    @Test
    fun settlesWithoutOvershoot() {
        val s = Spring(6.0)
        s.reset(0.0)
        var peak = 0.0
        var x = 0.0
        var t = 0.0
        while (t < 1.0) {
            x = s.step(100.0, 0.02)
            peak = maxOf(peak, x)
            t += 0.02
        }
        assertTrue(abs(x - 100.0) < 1.0)
        assertTrue(peak <= 100.0 + 1e-9)
    }

    @Test
    fun frameRateIndependent() {
        val fine = Spring(6.0).apply { reset(10.0) }
        val coarse = Spring(6.0).apply { reset(10.0) }
        var xFine = 0.0
        repeat(100) { xFine = fine.step(50.0, 0.01) }
        val xCoarse = coarse.step(50.0, 1.0)
        assertEquals(xFine, xCoarse, 1e-6)
    }
}

class PhoneFixtureTest {
    private fun fixtures(prefix: String) = File(CorePaths.core, "fixtures").listFiles()!!
        .filter { it.name.startsWith(prefix) && it.name.endsWith(".json") }
        .sortedBy { it.name }

    @Test
    fun phoneMappingFixturesReplay() {
        val pm = PhoneMapping(EffectConfig.DEFAULTS)
        val files = fixtures("phone-").filter { JSONObject(it.readText()).getString("platform") == "phone" }
        assertTrue("expected phone fixtures", files.size >= 2)
        for (file in files) {
            val fx = JSONObject(file.readText())
            val panel = fx.getString("panel")
            val tol = fx.getJSONObject("tolerance")
            val samples = fx.getJSONArray("samples")
            val byTime = HashMap<Double, Double>()
            for (i in 0 until samples.length()) {
                val s = samples.getJSONArray(i)
                byTime[s.getDouble(0)] = s.getDouble(1)
            }
            val expect = fx.getJSONArray("expect")
            for (i in 0 until expect.length()) {
                val e = expect.getJSONObject(i)
                val angle = byTime[e.getDouble("t")] ?: error("${file.name}: no sample at t=${e.getDouble("t")}")
                val progress = if (panel == "inner") pm.inner(angle) else pm.cover(angle)
                assertEquals("${file.name} t=${e.getDouble("t")} progress", e.getDouble("progress"), progress, tol.getDouble("progress"))
                assertEquals("${file.name} t=${e.getDouble("t")} tilt", e.getDouble("tilt"), pm.tilt(angle), tol.getDouble("deg"))
                assertEquals("${file.name} t=${e.getDouble("t")} settled", e.getBoolean("settled"), pm.isSettled(angle))
            }
        }
    }

    @Test
    fun detentFixturesReplay() {
        val files = fixtures("phone-").filter { JSONObject(it.readText()).getString("platform") == "phone-detent" }
        assertTrue("expected detent fixtures", files.size >= 2)
        for (file in files) {
            val fx = JSONObject(file.readText())
            val d = DetentDetector()
            val values = fx.getJSONArray("values")
            for (i in 0 until values.length()) d.feed(values.getDouble(i))
            assertEquals(file.name, fx.getBoolean("expectDetent"), d.isDetent)
        }
    }
}

File: windows/Hingewave.Core.Tests/CoreTests.cs
using System.Text.Json;
using Hingewave.Core;
using Xunit;

namespace Hingewave.Core.Tests;

/// <summary>Locates the repository's core/ directory from the test binary location.</summary>
internal static class CorePaths
{
    public static string Core
    {
        get
        {
            var dir = new DirectoryInfo(AppContext.BaseDirectory);
            for (var i = 0; i < 8 && dir != null; i++, dir = dir.Parent)
            {
                var candidate = Path.Combine(dir.FullName, "core", "effect.json");
                if (File.Exists(candidate)) return Path.Combine(dir.FullName, "core");
            }
            throw new DirectoryNotFoundException("core/effect.json not found above " + AppContext.BaseDirectory);
        }
    }
}

public class EffectConfigParityTests
{
    [Fact]
    public void DefaultsMatchEffectJson()
    {
        var fromFile = EffectConfig.Load(Path.Combine(CorePaths.Core, "effect.json"));
        Assert.Equal(EffectConfig.Defaults, fromFile);
    }
}

public class SpringTests
{
    [Fact]
    public void SettlesWithoutOvershoot()
    {
        var s = new Spring(6.0);
        s.Reset(0.0);
        double peak = 0.0, x = 0.0;
        for (var t = 0.0; t < 1.0; t += 0.02)
        {
            x = s.Step(100.0, 0.02).X;
            peak = Math.Max(peak, x);
        }
        Assert.True(Math.Abs(x - 100.0) < 1.0);
        Assert.True(peak <= 100.0 + 1e-9);
    }

    [Fact]
    public void FrameRateIndependent()
    {
        var fine = new Spring(6.0);
        var coarse = new Spring(6.0);
        fine.Reset(10.0);
        coarse.Reset(10.0);
        var xFine = 0.0;
        for (var i = 0; i < 100; i++) xFine = fine.Step(50.0, 0.01).X;
        var xCoarse = coarse.Step(50.0, 1.0).X;
        Assert.Equal(xCoarse, xFine, 6);
    }
}

public class FixtureTests
{
    private sealed record Expectation(double t, string state, double tilt, double progress, bool capture);
    private sealed record Tolerance(double deg, double progress);
    private sealed record LaptopFixture(string platform, double captureLatencyMs, bool reduceMotion, double[][] samples, Expectation[] expect, Tolerance tolerance);

    [Fact]
    public void AllLaptopFixturesReplay()
    {
        var files = Directory.GetFiles(Path.Combine(CorePaths.Core, "fixtures"), "laptop-*.json").OrderBy(f => f).ToArray();
        Assert.True(files.Length >= 4, "expected laptop fixtures in core/fixtures");
        var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true, IncludeFields = true };
        foreach (var file in files)
        {
            var fixture = JsonSerializer.Deserialize<LaptopFixture>(File.ReadAllText(file), options)!;
            Assert.Equal("laptop", fixture.platform);
            Replay(fixture, Path.GetFileName(file));
        }
    }

    private static void Replay(LaptopFixture fixture, string name)
    {
        var model = new LaptopMotionModel(EffectConfig.Defaults, fixture.reduceMotion);
        double? armedAt = null;
        var outputs = new Dictionary<double, MotionOutput>();
        var latency = fixture.captureLatencyMs / 1000.0;

        foreach (var sample in fixture.samples)
        {
            double t = sample[0], angle = sample[1];
            if (armedAt is double a && t >= a + latency)
            {
                model.CaptureReady(t);
                armedAt = null;
            }
            var output = model.Feed(t, angle);
            if (output.State == MotionState.Armed && armedAt == null && !model.CaptureSeen) armedAt = t;
            outputs[t] = output;
        }

        foreach (var e in fixture.expect)
        {
            Assert.True(outputs.TryGetValue(e.t, out var output), $"{name}: no output at t={e.t}");
            Assert.True(string.Equals(output.State.ToString(), e.state, StringComparison.OrdinalIgnoreCase), $"{name} t={e.t} state {output.State} != {e.state}");
            Assert.True(Math.Abs(output.Tilt - e.tilt) <= fixture.tolerance.deg, $"{name} t={e.t} tilt {output.Tilt} != {e.tilt}");
            Assert.True(Math.Abs(output.Progress - e.progress) <= fixture.tolerance.progress, $"{name} t={e.t} progress {output.Progress} != {e.progress}");
            Assert.True(output.Capture == e.capture, $"{name} t={e.t} capture");
        }
    }
}
